modules/migrations/gitlab.go

```diff
@@ -0,0 +1,524 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/migrations/base"
+	"code.gitea.io/gitea/modules/structs"
+
+	"github.com/xanzy/go-gitlab"
+)
+
+var (
+	_ base.Downloader        = &GitlabDownloader{}
+	_ base.DownloaderFactory = &GitlabDownloaderFactory{}
+)
+
+func init() {
+	RegisterDownloaderFactory(&GitlabDownloaderFactory{})
+}
+
+// GitlabDownloaderFactory defines a gitlab downloader factory
+type GitlabDownloaderFactory struct {
+}
+
+// Match returns ture if the migration remote URL matched this downloader factory
+func (f *GitlabDownloaderFactory) Match(opts base.MigrateOptions) (bool, error) {
+	var matched bool
+
+	u, err := url.Parse(opts.CloneAddr)
+	if err != nil {
+		return false, err
+	}
+	if strings.EqualFold(u.Host, "gitlab.com") && opts.AuthUsername != "" {
+		matched = true
+	}
+
+	return matched, nil
+}
+
+// New returns a Downloader related to this factory according MigrateOptions
+func (f *GitlabDownloaderFactory) New(opts base.MigrateOptions) (base.Downloader, error) {
+	u, err := url.Parse(opts.CloneAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	baseURL := u.Scheme + "://" + u.Host
+	repoNameSpace := strings.TrimPrefix(u.Path, "/")
+
+	log.Trace("Create gitlab downloader. BaseURL: %s RepoName: %s", baseURL, repoNameSpace)
+
+	return NewGitlabDownloader(baseURL, repoNameSpace, opts.AuthUsername, opts.AuthPassword), nil
+}
+
+// GitServiceType returns the type of git service
+func (f *GitlabDownloaderFactory) GitServiceType() structs.GitServiceType {
+	return structs.GitlabService
+}
+
+// GitlabDownloader implements a Downloader interface to get repository informations
+// from gitlab via go-gitlab
+// - issueCount is incremented in GetIssues() to ensure PR and Issue numbers do not overlap,
+// because Gitlab has individual Issue and Pull Request numbers.
+// - issueSeen, working alongside issueCount, is checked in GetComments() to see whether we
+// need to fetch the Issue or PR comments, as Gitlab stores them separately.
+type GitlabDownloader struct {
+	ctx             context.Context
+	client          *gitlab.Client
+	repoID          int
+	repoName        string
+	issueCount      int64
+	fetchPRcomments bool
+}
+
+// NewGitlabDownloader creates a gitlab Downloader via gitlab API
+//   Use either a username/password, personal token entered into the password field, or anonymous/public access
+//   Note: Public access only allows very basic access
+func NewGitlabDownloader(baseURL, repoPath, username, password string) *GitlabDownloader {
+	var client *http.Client
+	var gitlabClient *gitlab.Client
+	var err error
+	if username != "" {
```

Comment:
On github downloader, if password is empty, it think username is token. All the UI think that. So here will not work when fill token on password.